Improved time series annotation interface

samueljackson92 opened the pull request:
This makes a number of changes to improve the time series annotation interface.

1. Adding better zooming functionality. User can now toggle "edit mode"/ "zoom mode" and scroll zoom on traces
2. Annotator settings are persistent between shots
3. Batch select, edit, and deletion of annotations has been implemented
4. Import and export annotation buttons have been added
5. Previous/next navigation buttons have been added.

samueljackson92 committed:
Add improved zooming

samueljackson92 committed:
Add ability to import & export annotations

samueljackson92 committed:
Add next and previous navigation bar

samueljackson92 committed:
Fix performance of data trace zoom

samueljackson92 committed:
Avoid re-rendering on annotation replot

samueljackson92 committed:
Fix unzoom when applying tools

samueljackson92 committed:
Fix zoom rendering

samueljackson92 committed:
Add ability multi-select zone annotations

samueljackson92 committed:
Add small offset to min/max scaling

samueljackson92 committed:
tweak control keys for editing

samueljackson92 committed:
Tidy up time series a little bit

samueljackson92 committed:
Rename view

samueljackson92 committed:
Fix prettier formatting

samueljackson92 committed:
Update build

samueljackson92 committed:
Fix a few eslint issues

samueljackson92 committed:
Fix CI issues

samueljackson92 committed:
Add more classes properties

samueljackson92 committed:
Fix unit test

samueljackson92 committed:
Fix prettier formatting

samueljackson92 committed:
Merge branch 'main' into samueljackson92/improved-time-series-interface

wk9874 commented on toktagger/api/core/annotators.py:
```diff
@@ -234,7 +234,7 @@ def predict(self, data: MultiVariateTimeSeriesData) -> list[TimeRegion]:
             if peak_time >= tmin and peak_time <= tmax:
                 region = TimeRegion(
-                    label="Peak",
+                    label="Unknown",
```
Why have these changed to unknown?

samueljackson92 replied:
Label is the type e.g. `UFO`, `ELM`, `IRE` etc. We now have a default option of `Unknown`.

samueljackson92 committed:
Merge branch 'main' into samueljackson92/improved-time-series-interface

wk9874 commented on toktagger/api/core/query_strategy.py:
```diff
@@ -8,28 +10,49 @@
-        return self.samples.pop(index)
+        return self.samples[previous_index]
+
+    def _get_matching_sample(self, current_sample_id: Optional[str]) -> int:
```
Why is `current_sample_id` optional here, but required in higher level functions which call this?

wk9874 commented on toktagger/api/core/query_strategy.py:
```diff
@@ -38,38 +61,52 @@ class SequentialQueryStrategy(QueryStrategy):
+        idx = np.arange(len(self.samples))
+        rng = np.random.default_rng(seed=seed)
+        rng.shuffle(idx)
+        self.samples = [self.samples[i] for i in idx]
```
Wouldnt it be easier to just do `random.shuffle(self.samples)`?

samueljackson92 replied:
Random needs to be seeded so we keep the same (random) ordering with each call to the API. Otherwise `next` and `previous` will not work as expected.

wk9874 commented on toktagger/api/core/query_strategy.py:
```diff
@@ -38,38 +61,52 @@ class SequentialQueryStrategy(QueryStrategy):
+    """
+
+    def __init__(self, samples, annotations, seed: int = 42):
+        super().__init__(samples, annotations)
```
This will already shuffle the samples since it inherits from RandomQueryStrategy, making the following if statement redundant

wk9874 commented on toktagger/api/routers/samples.py:
```diff
@@ -187,14 +188,53 @@ async def get_next_sample(
+    project_id: str = Path(
+        description="The project to return the previous sample from."
+    ),
+    current_sample_id: Optional[str] = Query(
```
This shouldnt be optional here as it is required for getting previous sample from query strategy

wk9874 commented on toktagger/api/routers/samples.py:
```diff
@@ -187,14 +188,53 @@ async def get_next_sample(
+    db_client = request.app.state.db_client
+    project = await utils.get_project(db_client, project_id)
+    samples = await utils.get_samples(db_client, project_id)
+    query_strategy = QUERY_STRATEGIES[project.query_strategy](samples)
```
Since this is reinitialized each time, the 'previous' button may be unreliable:

* If new samples are added during a labelling sessions with Random query strategy, and then someone presses previous, the order of the shuffled samples will be different, and so next will send them to a random sample

* If new annotations are added, eg by a model finishing predictions, and someone presses next with Uncertainty QS, the order of samples by uncertainty may have changed due to new annotations, and so itll send them to a new pulse, not the previous one they were looking at

wk9874 replied:
Think this is still unresolved, but happy for it to just be an edge case we accept might cause issues since I can't think of a way to recitify this atm...

samueljackson92 replied:
I think it's a little better because you can at least cycle through every sample. However, I think it's still an issue. I think we should fix this by having the concept of a "validated" or "visited" sample and properly tracking state. That's a bigger task than we can fit into this PR, so I think it's a job for another issue.

I created: https://github.com/ukaea/toktagger/issues/153 for this

wk9874 commented on toktagger/api/routers/samples.py:
```diff
@@ -187,14 +188,53 @@ async def get_next_sample(
-        sample = data_pool.query_strategy.get_next_sample()
+        sample = query_strategy.get_previous_sample(current_sample_id)
     except RuntimeError:
         raise HTTPException(status_code=204, detail="No more samples available!")
```
Error message should be more like 'no previous sample available'

wk9874 commented on toktagger/api/schemas/annotations.py:
```diff
@@ -9,6 +9,8 @@ class AnnotationIn(ConfiguredModel):
     created_by: AnnotatorTypes
     validated: bool = False
     uncertainty: Optional[float] = None
+    sample_id: Optional[str] = None
```
Why did these move into AnnotationIn? The endpoint should fill these in rather then them be required in the actual Annotation's schema

samueljackson92 replied:
What about round-tripping a import/export. If I export then my annotations file contains these properties, if I then try and import it again it will break if we don't allow this?

wk9874 replied:
Wouldn't it make more sense to have the export to be in form:

```
{
    sample_id: [annotations]
}
```

And then if you import them again via the UI itd use the batch update endpoint?

samueljackson92 replied:
I really don't like this I would much rather have a flat list of annotations which can be easily imported into a pandas dataframe or similar

wk9874 commented on toktagger/api/crud/utils.py:
```diff
@@ -252,3 +253,37 @@ async def get_sample_summary(
+    for sample_id, sample_annotations in sample_groups.items():
+        sample_obj_id = convert_to_objectid(sample_id, "samples")
+
+        if not await db_client.get_document_by_id("samples", sample_obj_id):
```
Do we want this to fail on the first invalid sample ID? Or continue trying all others, then throw one error at the end with a list of any invalid sample IDs?

samueljackson92 replied:
I think fail here. Something has gone wrong at this point.

wk9874 commented on toktagger/api/routers/annotations.py:
```diff
@@ -67,6 +67,28 @@ async def get_all_annotations(
+)
+async def import_annotations(
+    request: Request,
+    annotations: list[AnnotationTypes],
```
Ahh now I see why you moved `sample_id` into `AnnotationIn` instead of keeping it in `Annotation` only - you need a sample ID attached to each annotation...

What I did instead in the models stuff was add a new schema:

```
class AnnotationBatchItem(ConfiguredModel):
    sample_id: str
    annotations: list[AnnotationTypes]
```

Your endpoint would then expect (I assume):
```
annotations: list[AnnotationBatchItem]
```

I personally think this is better since it still means you can define an Annotation without having to care about the project or sample ID, which is consistent with all our other schemas

samueljackson92 replied:
Ah ok. I think this makes sense.

samueljackson92 replied:
I don't see how this will work when importing annotations for multiple samples?

samueljackson92 replied:
I think you'd have to have a list of batched items and similarly for exporting. But I don't like this, thinking about how I want to use this *as a user*, I just want a flat list of annotations that I can load into a pandas dataframe...

With the approach I have implemented you still don't have to provide `sample_id` and `project_id` when creating annotations, but if you do have it then it can be optionally attached.

wk9874 commented on toktagger/ui/src/app/components/tools/toolbar.tsx:
```diff
@@ -34,36 +31,39 @@ import {
 ) {
+  // user has validated the annotations, so set created_by to "manual"
+  const updatedAnnotations = annotations.map((annotation: Annotation) => {
+    annotation.created_by = "manual";
```
Do you also need to set validated = True?

wk9874 commented on toktagger/ui/src/app/core.tsx:
```diff
@@ -102,3 +102,99 @@ export const deleteProject = async (project_id: string) => {
+  }
+}
+
+export async function saveAnnotations(
```
how does this work without being provided a sample ID to save to?

samueljackson92 replied:
sample id's are in the file.

wk9874 commented on tests/api/core/test_query_strategy.py:
```diff
@@ -86,13 +88,13 @@ def test_uncertainty_strategy(samples, annotations):
-    next_sample = strategy.get_next_sample()
-    assert next_sample.id == "sample_1"
-    next_sample = strategy.get_next_sample()
     assert next_sample.id == "sample_2"
```
Why have these reversed order? Indicates uncertainty query strat isnt working properly, should be giving us most uncertain first (annotation 5, linked to sample 4)

wk9874 replied:
Yeah this test is wrong now, the correct order (highest uncertainty first) is sample 4, then 1, then 2, then fallback to random selection (so sample 3 as that's the only one left)

See comment above about an extra reverse() in the UncertaintyQueryStrategy for what might cause this

wk9874 commented on tests/api/core/test_query_strategy.py:
```diff
@@ -54,27 +52,31 @@ def test_sequential_strategy(samples, annotations):
+    assert random_samples != samples
 
 
 def test_uncertainty_strategy(samples, annotations):
```
Should add 'preivous sample' versions of these tests for each query strategy

wk9874 commented on toktagger/api/routers/samples.py:
```diff
@@ -187,14 +188,53 @@ async def get_next_sample(
+    # And the /annotation endpoint to get initial prediction (if available)
+    db_client = request.app.state.db_client
+    project = await utils.get_project(db_client, project_id)
+    samples = await utils.get_samples(db_client, project_id)
```
By default this will return samples in the order in which they were created. Do we instead potentially want it to sort by shot ID?

Currently with the Local ELM demo project, when I press previous / next it unexpectedly goes in the 'wrong' order - instinctively I would expect it to go in shot ID order, but it does the reverse (because the samples were created highest shot ID first)

wk9874 replied:
Or split into two Query Strategies? One which is sequential by shot ID, one which is sequential by time created?

samueljackson92 replied:
Keep it simple: sort by shot id.

wk9874 commented on toktagger/ui/src/app/components/tools/export.tsx:
```diff
@@ -0,0 +1,81 @@
+
+  const handleExport = () => {
+    if (exportOption === "All") {
+      exportAnnotations(project);
```
The annotations returned differs between 'all' versus 'current:

With 'all':
```
  {
    "timestamp": "2025-11-21T12:07:03.399789",
    "label": "Sawtooth",
    "created_by": "manual",
    "validated": false,
    "uncertainty": 1,
    "sample_id": "692052ff8d2504a1af2843a5",
    "project_id": "692052ff8d2504a1af2843a3",
    "_id": "692055e78d2504a1af2843ae",
    "type": "time_region",
    "time_min": 0.4394439237225475,
    "time_max": 0.4589800451965979
  },
```

With current:
```
  {
    "created_by": "manual",
    "type": "time_region",
    "time_min": 0.4394439237225475,
    "time_max": 0.4589800451965979,
    "label": "Sawtooth"
  }
```

wk9874 replied:
Ok now theyre consistent, but project_id and sample_id are filled with nulls when exporting only the current sample:
```
[
  {
    "created_by": "manual",
    "type": "time_region",
    "project_id": null,
    "sample_id": null,
    "validated": false,
    "uncertainty": 1,
    "time_min": 5.311458196992046,
    "time_max": 5.902732015804313,
    "label": "ELM"
  },
  {
    "project_id": null,
    "sample_id": null,
    "validated": false,
    "uncertainty": 1,
    "created_by": "disruption_cnn",
    "type": "time_point",
    "time": 7.837500307692308,
    "label": "Disruption"
  },
  {
    "project_id": null,
    "sample_id": null,
    "validated": false,
    "uncertainty": 1,
    "created_by": "manual",
    "type": "time_point",
    "time": 9.026386117988075,
    "label": "Disruption"
  }
]
```

wk9874 commented on toktagger/ui/src/app/components/tools/export.tsx:
```diff
@@ -0,0 +1,81 @@
+  sample,
+  current_annotations,
+}: ExportToolInfo) {
+  const exportItems = ["All", "Current Sample"].map((item) => ({
```
I'm not sure it makes sense to give the option of 'all' or 'current' on the sample screen? Shouldnt we have two places where you can export annotations:

* Sample screen, exports that one sample only
* Project screen (table with list of samples), exports all samples

?

(could be a later PR, we should probably add a toolbar to the projects screen to add things like this, along with eg the models training / predict stuff)

wk9874 commented on toktagger/ui/src/app/components/tools/import.tsx:
```diff
@@ -0,0 +1,51 @@
+        <Flex justifyContent="center" alignItems="center">
+          <Button variant="primary">
+            <Import />
+            <Text>Import Annotations</Text>
```
Tried uploading the same form of data downloaded from export, eg:

```
[
  {
    "created_by": "manual",
    "type": "time_region",
    "time_min": 0.4394439237225475,
    "time_max": 0.4589800451965979,
    "label": "Sawtooth"
  }
]
```

And didnt seem to work:
```
INFO:     127.0.0.1:55906 - "PUT /projects/692052ff8d2504a1af2843a3/annotations HTTP/1.1" 404 Not Found
```

wk9874 commented on toktagger/ui/src/app/components/tools/import.tsx:
```diff
@@ -0,0 +1,51 @@
+        <Flex justifyContent="center" alignItems="center">
+          <Button variant="primary">
+            <Import />
+            <Text>Import Annotations</Text>
```
Again may be good to split this out into two:

* Import annotations on sample screen - just a list of annotations for this sample, dont need to provide the sample ID
* Import annotations on project screen: Annotations for a range of samples, must provide sample IDs in some way

samueljackson92 replied:
I think it's nice to have the option of either all or just this sample on the sample page?

wk9874 replied:
I think for this, I would have the 'Import Annotations' button on the sample page not require sample / project IDs in the schema which is uploaded. You could have it still accept both I guess, but:


1. If the annotations uploaded by the user dont contain a project ID and/or sample ID, automatically fill it with the proejct and sample currently being looked at
2. If the annotations do contain a project or sample ID, these take priority over the page the user is currently looking at (?)

wk9874 replied:
Think this hasn't been addressed - if I upload an annotation via the 'Import Annotations' button on the sample page (not the samples table) like this:

```
[
  {
    "label": "Disruption",
    "created_by": "manual",
    "validated": true,
    "uncertainty": 0,
    "type": "time_point",
    "time": 8.4
  }
]
```
I think it should fill the shot ID, sample ID, project ID automatically for me based on the sample I am currently looking at

wk9874 commented on toktagger/ui/src/app/components/tools/nav.tsx:
```diff
@@ -0,0 +1,194 @@
+  const navigate = useNavigate();
+
+  const moveNextShot = useCallback(async () => {
+    await saveSampleAnnotations(project_id, sample_id, annotations);
```
Im not convinced we should silently autosave when someone presses previous/next buttons without pressing save... what if someone is just flicking between different shots to see what the auto annotator tools look like? Theyll unknowingly be adding 'validated' annotations into the database and marking shots as validated

Maybe it could popup with 'save shots'? With a yes/no, and a 'set default / dont ask again' option?

samueljackson92 replied:
Improvement for another PR...

wk9874 replied:
Will leave this as a comment on #54

wk9874 commented on toktagger/ui/src/app/components/plots/time-series.tsx:
```diff
@@ -65,9 +83,15 @@ export const TimeSeries = ({
     setUpdateTools((current) => (current + 1) % 100);
   };
 
+  if (!disableToolingInteraction) {
```
How do i enable zoom mode on the UI?

samueljackson92 replied:
Hold down "Alt" and scroll to scroll zoom. We should have a page showing keyboard shortcuts.... Another PR...

wk9874 replied:
Raised #149

samueljackson92 committed:
Checkpoint work

samueljackson92 committed:
Fix zooming issues

samueljackson92 committed:
Tidy-up

samueljackson92 committed:
Fix navigation

samueljackson92 committed:
refactoring time series annotation

samueljackson92 committed:
Fix types

samueljackson92 committed:
Clean out old code. Make Vspans great again

samueljackson92 committed:
Fixing some knock on spectrogram view issues

samueljackson92 committed:
fix a few issue with the spec view

samueljackson92 committed:
Fix issue with vspan annotation colors

samueljackson92 committed:
Clean up project structure

samueljackson92 committed:
Fix random ordering

samueljackson92 committed:
Fix unit tests

samueljackson92 committed:
More updates:
  - Add time selection range to UI
  - Add error handling page
  - Add/update favicon

samueljackson92 committed:
Add loading progress support

samueljackson92 committed:
Update with sample context

samueljackson92 committed:
Code refactoring

samueljackson92 committed:
Add logo

samueljackson92 committed:
Add more work on context

samueljackson92 committed:
Update import/export buttons

samueljackson92 committed:
add tests for import/export

samueljackson92 committed:
Add time resolution option

samueljackson92 committed:
Add documentation for UI

samueljackson92 committed:
Merge branch 'main' into samueljackson92/improved-time-series-interface

samueljackson92 committed:
Fix a few more issues after merge

Commit:
chore: update build output [skip ci]

samueljackson92 committed:
Fix some unit tests

samueljackson92 committed:
Fix unit tests

samueljackson92 committed:
Prettier linting

samueljackson92 committed:
Fix eslint issues

samueljackson92 committed:
Update build

samueljackson92 committed:
Prettier linting

Commit:
chore: update build output [skip ci]

wk9874 commented on toktagger/api/core/query_strategy.py:
```diff
@@ -11,29 +13,56 @@
-        return self.samples.pop(index)
+        return self.samples[previous_index]
+
+    def _get_matching_sample(self, current_sample_id: Optional[str]) -> int:
```
Picky comment: I think this method should be defined before `get_next_sample` and `get_previous_sample`, since those methods use this one

Also the typehint on `current_sample_id` is wrong here since it is a required param

wk9874 commented on toktagger/api/core/query_strategy.py:
```diff
@@ -42,40 +71,91 @@ class SequentialQueryStrategy(QueryStrategy):
+        rng = np.random.default_rng(seed=seed)
+        self.samples = rng.permutation(self.samples)
+
+    def get_next_sample(self, current_sample_id: Optional[str] = None) -> Sample:
```
This is redundant since it should be inherited from QueryStrategy (base method is identical to this as far as I can tell)

wk9874 commented on toktagger/api/core/query_strategy.py:
```diff
@@ -42,40 +71,91 @@ class SequentialQueryStrategy(QueryStrategy):
-        return self.samples.pop(0)
+        return self.samples[next_index]
 
+    def get_previous_sample(self, current_sample_id: Optional[str] = None) -> Sample:
```
Same here - should be inherited from QueryStrategy

wk9874 commented on toktagger/api/core/query_strategy.py:
```diff
@@ -42,40 +71,91 @@ class SequentialQueryStrategy(QueryStrategy):
-    def get_next_sample(self) -> Sample:
-        if len(self.samples) == 0:
-            raise RuntimeError("No more samples to label!")
+    def __init__(self, samples, annotations, seed: int = 42):
```
Should add type hints here

wk9874 replied:
Also `annotations` should be optional, like all other query strategies

wk9874 commented on toktagger/api/core/query_strategy.py:
```diff
@@ -42,40 +71,91 @@ class SequentialQueryStrategy(QueryStrategy):
+                key=lambda sample: sample_ids.index(sample.id)
+                if sample.id in sample_ids
+                else -1,
+                reverse=True,
```
I think we might have one too many `reverse=True`'s in here. `self.annotations` returns in reverse order already, so highest uncertainty first. If we then reverse it again here, wont we get them in uncertainty lowest -> highest order?

wk9874 commented on tests/api/routers/test_annotations.py:
```diff
@@ -312,6 +387,7 @@ async def test_batch_update_annotations(api_client, setup_db, db_client):
         sort_by="time",
         sort_direction="ascending",
     )
+    print(annotations_sample_1)
```
Remove

wk9874 replied:
Remove

wk9874 commented on tests/api/routers/test_models.py:
```diff
@@ -264,6 +264,7 @@ async def test_model_get_sample_prediction_wrong_sample(
     get_response = await api_client.get(
         f"/projects/{setup_model_db['project_id']}/samples/{setup_model_db['sample_ids'][-2]}/models/mock_disruption_cnn/predict/{task_id}"
     )
+    print(get_response.json())
```
Remove

wk9874 replied:
Remove

wk9874 commented on tests/api/routers/test_projects.py:
```diff
@@ -37,13 +39,13 @@ async def test_get_all_projects_sortby(api_client, setup_db):
     # Should sort alphabetically by task
     # Sorts by case first (all uppers before any lowers)
-    # So ELM (project 1), then UFO (project 2), then disruption (project 0)
+    # So time series (project 1), then video (project 2), then time series (project 0)
```
Update this comment to be correct

Also might want to either change the sort_by parameter, or change one of the projects to be a spectogram task? Because currently its arbitrary which of projects 0 or 1 it returns first since they have the same task

wk9874 replied:
Update this comment to be correct

Also might want to either change the sort_by parameter, or change one of the projects to be a spectogram task? Because currently its arbitrary which of projects 0 or 1 it returns first since they have the same task

wk9874 commented on toktagger/api/core/data_loaders.py:
```diff
@@ -38,7 +45,9 @@ def sample_data_type(cls) -> Type[ShotData | FileData | TimeSeriesFileData]:
     def get_sample(
-        self, shot_id: int, sample_data: ShotData | FileData | TimeSeriesFileData
+        self,
+        sample: ShotData | FileData | TimeSeriesFileData,
```
This typing is now wrong

wk9874 replied:
This typing is now wrong

wk9874 commented on toktagger/api/core/data_loaders.py:
```diff
@@ -97,7 +106,14 @@ def sample_data_type(self) -> Type[FileData]:
 
     @pydantic.validate_call
-    def get_sample(self, shot_id: int, sample_data: FileData) -> ImageData:
+    def get_sample(self, sample: Sample, **kwargs) -> ImageData:
```
Was this intentionally changed back to the old format? I split the arguments out into `shot_id` and `sample_data` so that pydantic could automatically validate the sample data model whenever this method was called

wk9874 replied:
(not essential anymore though, I think i initially did that when the loader registry relied on the type of `sample_data` to get the model schema, but realised that was a bad idea and added the `sample_data_type` classmethod instead)

wk9874 replied:
^

wk9874 commented on toktagger/api/core/data_loaders.py:
```diff
@@ -97,7 +106,14 @@ def sample_data_type(self) -> Type[FileData]:
+    def get_sample(self, sample: Sample, **kwargs) -> ImageData:
+        sample_data: FileData = sample.data
+
+        if not isinstance(sample_data, FileData):
```
So eg splitting `sample_data` out in the arguments to the data loader makes this check redundant, as pydantic will automatically validate it on method call

wk9874 commented on toktagger/api/schemas/annotations.py:
```diff
@@ -99,21 +97,22 @@ class ModelAnnotation(AnnotationIn):
 
 AnnotationTypeAdapter = TypeAdapter(AnnotationTypes)
 AnnotationOutTypeAdapter = TypeAdapter(AnnotationOutTypes)
+AnnotationBatchInputTypeAdapter = TypeAdapter(AnnotationBatchInputTypes)
```
This is now redundant as it is identical to AnnotationOutTypeAdapter

wk9874 commented on toktagger/api/schemas/annotations.py:
```diff
@@ -99,21 +97,22 @@ class ModelAnnotation(AnnotationIn):
-    sample_id: str
-    annotations: list[AnnotationTypes]
-
+AnnotationBatchInputTypes = AnnotationOutTypes
```
Makes this redundant

wk9874 commented on toktagger/api/routers/annotations.py:
```diff
@@ -69,76 +68,58 @@ async def get_all_annotations(
-async def delete_all_annotations(
+async def import_annotations(
     request: Request,
+    annotations: list[AnnotationBatchInputTypes],
```
Currently `import_annotations` and `update_annotations` in the annotations router page both expect an `AnnotationBatchInputTypes` list, which (might? since they're optional) contain the `sample_id` and `project_id` within the Annotation.

`import_annotations` needs to contain the sample ID in some way alongside the annotations, which was why I previous had:

```
class AnnotationBatchItem(ConfiguredModel):
    sample_id: str
    annotations: list[AnnotationTypes]
```
and `import_annotations` expects a list of those. It should not need `project_id` to be provided in the annotation itself

wk9874 commented on toktagger/api/routers/annotations.py:
```diff
@@ -200,13 +181,13 @@ async def get_annotations(
 async def update_annotations(
     request: Request,
-    annotations: list[AnnotationTypes],
+    annotations: list[AnnotationBatchInputTypes],
```
Meanwhile `update_annotations` shouldn't have a project_id or sample_id provided within the annotation. What if the values provided in the annotation arent the same as those in the path of the API request? The user might see unexpected behaviour here


Generally, this is why I'm going back to `AnnotationIn` not having `sample_id` or `project_id` as optional parameters. It means that we know exactly what format and what parameters are expected by anything which accepts an `AnnotationType`. Currently we have the confusing situation where two different endpoints accept the same type, but get the sample IDs from different places

wk9874 commented on toktagger/api/schemas/annotations.py:
```diff
@@ -70,26 +69,25 @@ class VideoBoundingBox(BoundingBox):
-    sample_id: str
+class AnnotationOut(AnnotationIn):
+    id: Optional[str] = Field(None, alias="_id")
+    project_id: Optional[str] = None
```
Why are these optional now?

wk9874 replied:
Why are these optional now?

wk9874 commented on toktagger/ui/src/app/projects/project_id/page.tsx:
```diff
@@ -175,15 +197,21 @@ export default function ProjectView() {
-              />
+              <Flex gap="size-100" alignItems="center" justifyContent="start">
+                <AddSamplesEditor project={project} onModify={refreshSamples} />
+                <SearchField
```
In terms of page layout, the search bar on the Projects and Samples page should be in the same location:

<img width="1571" height="729" alt="Image" src="https://github.com/user-attachments/assets/4ae907eb-1755-4e8d-9c1f-6a0c6b094d8a" />

<img width="1569" height="716" alt="Image" src="https://github.com/user-attachments/assets/f25d7380-6e45-4a42-a017-bd15319fc36e" />

wk9874 replied:
In terms of page layout, the search bar on the Projects and Samples page should be in the same location:

<img width="1571" height="729" alt="Image" src="https://github.com/user-attachments/assets/4ae907eb-1755-4e8d-9c1f-6a0c6b094d8a" />

<img width="1569" height="716" alt="Image" src="https://github.com/user-attachments/assets/f25d7380-6e45-4a42-a017-bd15319fc36e" />

wk9874 commented on scripts/setup.py:
```diff
@@ -4,12 +4,21 @@
         "query_strategy": query_strategy,
         "data_loader": data_loader,
+        "time_min": -0.1,
+        "time_max": 0.8,
```
Need to expose these time settings as options to `create_project`, and then add `create_project(..., time_max=times[-1])` into the `create_mock_data.py` script

wk9874 commented on toktagger/api/routers/samples.py:
```diff
@@ -170,77 +157,84 @@ async def add_samples(
-
-    query_strategy = QUERY_STRATEGIES[project.query_strategy](samples, annotations)
+    samples = await utils.get_samples(db_client, project_id)
+    query_strategy = QUERY_STRATEGIES[project.query_strategy](samples)
```
This never passes in annotations to the Query Strategy

wk9874 commented on toktagger/api/routers/samples.py:
```diff
@@ -170,77 +157,84 @@ async def add_samples(
-    await utils.get_project(db_client, project_id)
+    project = await utils.get_project(db_client, project_id)
+    samples = await utils.get_samples(db_client, project_id)
+    query_strategy = QUERY_STRATEGIES[project.query_strategy](samples)
```
This never passes annotations into the query strategy, so Uncertainty based QS wont work

wk9874 commented on toktagger/ui/src/app/components/tools/export.tsx:
```diff
@@ -0,0 +1,81 @@
+    <div className="m-4">
+      <Flex direction="column">
+        <Flex direction="row" alignItems="center" gap="size-100">
+          <ComboBox
```
Personally I dont even think we should provide the option here... If you want to export all annotations for a project, do it from the project screen. If you want just a specific sample, do it from the sample screen

wk9874 commented on toktagger/ui/src/app/time_series/components/time-series.tsx:
```diff
@@ -0,0 +1,178 @@
+} from "@/app/utils";
+import { useEffect, useMemo, useState } from "react";
+
+const zoneCategories: Category[] = [
```
Where have eg `RampUp`, `FlatTop` etc gone?

wk9874 commented on toktagger/ui/src/app/time_series/components/time_series.tsx:
```diff
@@ -0,0 +1,178 @@
+            onModifyZone={updateZones}
+          >
+            <TimeSeries
+              plotId="ELMs"
```
Should this still reference ELMs?

wk9874 commented on toktagger/ui/src/app/projects/project_id/samples/sample_id/page.tsx:
```diff
@@ -63,67 +60,73 @@ type SampleViewInfo = {
+  >(null);
+
+  useEffect(() => {
+    if (project.task == "disruption") {
```
This 'disruption' task no longer exists

wk9874 replied:
In fact none of these project tasks exist...

samueljackson92 replied:
Yup gone. Diff was out of date...

wk9874 commented on toktagger/ui/src/app/components/providers/vpsan-provider.tsx:
```diff
@@ -69,8 +69,9 @@ export const VSpanProvider = ({
   };
 
-  const handleDelete = (input: unknown) => {
+  const handleVSpanDelete = (input: unknown) => {
```
What is input here?

wk9874 replied:
What is input here?

wk9874 commented on toktagger/ui/src/app/components/providers/vpsan-provider.tsx:
```diff
@@ -69,8 +69,9 @@ export const VSpanProvider = ({
-  const handleDelete = (input: unknown) => {
+  const handleVSpanDelete = (input: unknown) => {
     spans.current = spans.current.filter((span) => span !== input);
+    spans.current = spans.current.filter((span) => !span.selected);
```
Do you intend to do two updates to this here?

wk9874 replied:
Do you intend to do two updates to this here?

wk9874 commented on toktagger/ui/src/app/components/providers/vpsan-provider.tsx:
```diff
@@ -193,6 +199,17 @@ export const VSpanProvider = ({
 
+  // Delete selected zones on Delete/Backspace keypress
+  document.addEventListener("keydown", (e) => {
+    if (e.key === "Delete" || e.key == "Backspace") {
```
How do i select a zone for this to work? For me either clicking, or clicking and holding, and then pressing delete or backspace does nothing...

wk9874 commented on toktagger/ui/src/app/components/providers/zone-provider.tsx:
```diff
@@ -209,6 +215,19 @@ export const ZoneProvider = ({
+
+  // Delete selected zones on Delete/Backspace keypress
+  document.addEventListener("keydown", (e) => {
+    if (e.key === "Delete" || e.key == "Backspace") {
```
Same here, how do i use this?

wk9874 commented on toktagger/ui/src/app/components/providers/annotation-provider.tsx:
```diff
@@ -73,7 +73,7 @@ export const ContextMenuProvider = ({
   useEffect(() => {
     const disableInteraction = (event: KeyboardEvent) => {
-      if (event.key === "Shift" || event.key === "Control") {
+      if (event.key === "Alt") {
```
its not clear to me that Alt is doing anything?

samueljackson92 replied:
Have you tried zooming while hovering over an annotation?

wk9874 commented on toktagger/ui/src/app/components/tools/nav.tsx:
```diff
@@ -0,0 +1,312 @@
+async function getNextSample(project_id: string, current_sample_id: string) {
+  const NEXT_URL = `${BACKEND_API_URL}/projects/${project_id}/samples/next?current_sample_id=${current_sample_id}`;
+  const sampleResult = await fetch(NEXT_URL);
+  if (sampleResult.status === 204) {
```
Should catch other errors and display to user

wk9874 commented on toktagger/ui/src/app/components/tools/nav.tsx:
```diff
@@ -0,0 +1,312 @@
+) {
+  const PREVIOUS_URL = `${BACKEND_API_URL}/projects/${project_id}/samples/previous?current_sample_id=${current_sample_id}`;
+  const sampleResult = await fetch(PREVIOUS_URL);
+  if (sampleResult.status === 204) {
```
Same here, should catch other errors too and display error to user

wk9874 commented on toktagger/ui/src/app/components/tools/nav.tsx:
```diff
@@ -0,0 +1,312 @@
+
+  return (
+    <View marginStart="size-100">
+      <ActionButton onPress={moveNextShot}>
```
Should add an appropriate aria-label to this so that it can be identified by screen readers (and Playwright for end to end tests!)

wk9874 commented on toktagger/ui/src/app/components/tools/nav.tsx:
```diff
@@ -0,0 +1,312 @@
+
+  return (
+    <View marginStart="size-100">
+      <ActionButton onPress={movePreviousShot}>
```
Again, aria-label here

wk9874 replied:
(Note that for most elements React / React Spectrum handle this for us, and the text label automatically becomes the aria label. But anywhere that there are buttons with an icon only (no text), we should add a text aria-label to describe what the button is)

wk9874 commented on toktagger/ui/src/app/components/tools/nav.tsx:
```diff
@@ -0,0 +1,312 @@
+          isSelected={validateOnNavigate}
+          onChange={setValidateOnNavigate}
+        >
+          Validate on Navigate
```
Should this just change to 'Save on Navigate'? 'Validate' is potentially a confusing term for users

wk9874 replied:
I think more generally - what do we want this to do? Currently it will always save (And hence overwrite any preexisting) annotations when you navigate, its just they wont be marked as validated. I think personally if this box is not ticked, it shouldnt save the annotations on navigation at all... But I leave it up to you :)

wk9874 commented on toktagger/ui/src/app/components/tools/vspans.tsx:
```diff
@@ -116,6 +124,20 @@ export const VSpans = ({ plotId, plotReady, forceUpdate }: ToolingProps) => {
+          vspan.x > selectedXRange[0] &&
+          vspan.x < selectedXRange[1]
+        ) {
+          vspan.selected = true;
```
Cant test this as I dont know how to select a Vspan or Zone... (I assume clicking on it? If so, this isnt working)

wk9874 replied:
Used box select and this works

wk9874 commented on toktagger/ui/src/app/time_series/components/time-series.tsx:
```diff
@@ -0,0 +1,218 @@
+import { useSample } from "@/app/contexts/SampleContext";
+import { Flex, View } from "@adobe/react-spectrum";
+
+const zoneCategories: Category[] = [
```
disruption things like 'RampUp', 'FlatTop' missing

wk9874 commented on toktagger/ui/src/app/time_series/components/time-series.tsx:
```diff
@@ -0,0 +1,218 @@
+    });
+
+    // Build yaxis layout object dynamically
+    const yAxesLayout = yAxisDomains.reduce(
```
Where are the axes titles set? I dont see a title when there is only one plot (for multiple plots the key is displayed which acts as the title)

<img width="1591" height="853" alt="Image" src="https://github.com/user-attachments/assets/52635290-ce4c-43fb-90c5-c2da03f0e9b2" />

wk9874 commented on toktagger/ui/src/app/core.tsx:
```diff
@@ -103,6 +109,121 @@ export const deleteProject = async (project_id: string) => {
+  project_id: string,
+  sample_id: string,
+  annotations: Annotation[],
+  validateOnSave: boolean = true,
```
This should probably be called validateOnNavigate

wk9874 commented on toktagger/ui/src/app/core.tsx:
```diff
@@ -103,6 +109,121 @@ export const deleteProject = async (project_id: string) => {
+    return annotation;
+  });
+
+  const ANNOTATIONS_URL = `${BACKEND_API_URL}/projects/${project_id}/samples/${sample_id}/annotations`;
```
I wouldve thought we just wouldnt save if validateOnNavigate=False?

wk9874 commented on toktagger/ui/src/app/time_series/components/time-series.tsx:
```diff
@@ -0,0 +1,218 @@
+              </TimeSeries>
+            </VSpanProvider>
+          </ZoneProvider>
+        </ContextMenuProvider>
```
This means we've lost the little table which used to appear at the bottom of the disruption view, right?

I'm wondering whether we want to still have the old tasks which exist on the backend, so that we can display different things under the banner of 'time-series'.

Eg, if task == 'disruption', dont add some of those tools to the toolbar as they wouldnt be useful, and render the table like before. Also limit which Vspans and Zones are available.

We could either on the backend or the frontend store which type of data it is (ie disruption = TimeSeries, ELM = TimeSeries, UFO = Image, etc)

samueljackson92 replied:
Maybe, I think for now keeping it generic is ok for now. We can bring more specific things later if required.

I also only removed the table because I think it needs more work, but actually I can quickly provide some updates here to improve it and re-add it. Here's how it looks now:

<img width="1579" height="259" alt="image" src="https://github.com/user-attachments/assets/d5ca9b89-93a2-47de-87f5-e51b11347441" />

dependabot[bot] committed:
Bump react-router and react-router-dom in /toktagger/ui

Bumps [react-router](https://github.com/remix-run/react-router/tree/HEAD/packages/react-router) to 7.12.0 and updates ancestor dependency [react-router-dom](https://github.com/remix-run/react-router/tree/HEAD/packages/react-router-dom). These dependencies need to be updated together.


Updates `react-router` from 7.8.2 to 7.12.0
- [Release notes](https://github.com/remix-run/react-router/releases)
- [Changelog](https://github.com/remix-run/react-router/blob/main/packages/react-router/CHANGELOG.md)
- [Commits](https://github.com/remix-run/react-router/commits/react-router@7.12.0/packages/react-router)

Updates `react-router-dom` from 7.8.2 to 7.12.0
- [Release notes](https://github.com/remix-run/react-router/releases)
- [Changelog](https://github.com/remix-run/react-router/blob/main/packages/react-router-dom/CHANGELOG.md)
- [Commits](https://github.com/remix-run/react-router/commits/react-router-dom@7.12.0/packages/react-router-dom)

---
updated-dependencies:
- dependency-name: react-router
  dependency-version: 7.12.0
  dependency-type: indirect
- dependency-name: react-router-dom
  dependency-version: 7.12.0
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

samueljackson92 committed:
Fix issues from code review

samueljackson92 committed:
Fix small issue with deleting annotations

Commit:
Merge remote-tracking branch 'origin/main' into samueljackson92/improved-time-series-interface

Commit:
chore: update build output [skip ci]

samueljackson92 committed:
Update export to include missing sample names

Commit:
chore: update build output [skip ci]

samueljackson92 committed:
Remove redundant code

samueljackson92 committed:
Add axes labels to subplots

samueljackson92 committed:
Add annotations table - unused for now

samueljackson92 committed:
Make annotations nicer/more consistent

Commit:
chore: update build output [skip ci]

wk9874 committed:
Update create mock data script

samueljackson92 committed:
Remove more code duplication, remove console.log

Commit:
chore: update build output [skip ci]

samueljackson92 committed:
Fix tests, fix previous navigation

Commit:
chore: update build output [skip ci]

samueljackson92 committed:
rebuild project

samueljackson92 committed:
Merge branch 'main' into samueljackson92/improved-time-series-interface

Commit:
chore: update build output [skip ci]

wk9874 commented on toktagger/ui/src/app/projects/project_id/page.tsx:
```diff
@@ -175,15 +197,21 @@ export default function ProjectView() {
+                />
+              </Flex>
+              <Flex gap="size-100">
+                <ImportButton project_id={project_id} />
```
When importing annotations, I suspect from the user's POV it would be better if they could provide the annotations without the sample IDs, and instead just provide the shot IDs. The UI could then do the mapping between the two somehow?

But this could be raised as an issue and looked at in a future PR

samueljackson92 committed:
Update with code review changes

samueljackson92 committed:
Update logo and favicon

Commit:
Merge remote-tracking branch 'origin/main' into samueljackson92/improved-time-series-interface

Commit:
chore: update build output [skip ci]

wk9874 commented on toktagger/ui/src/app/components/plots/time-series.tsx:
```diff
@@ -244,6 +244,12 @@ export const TimeSeries = ({
       });
     };
 
+    // if (!disableToolingInteraction) {
```
Is this intentionally commented out?

wk9874 commented on toktagger/api/routers/annotations.py:
```diff
@@ -208,10 +208,15 @@ async def update_annotations(
+    # Set shot_id for each annotation
+    for annotation in annotations:
+        annotation.shot_id = sample.shot_id
+        print(annotation.shot_id)
```
Remove debug print statement

wk9874 commented on toktagger/ui/main.jsx:
```diff
@@ -2,6 +2,23 @@ import React from "react";
+}
+
+// Update on load
+updateFavicon();
```
I still see the old favicon when using `toktagger` to run the simulation - but this might be browser caching...

wk9874 replied:
Finallyyyy fixed

samueljackson92 committed:
Update from code review

wk9874 committed:
Favicon stuff

Commit:
chore: update build output [skip ci]

wk9874 committed:
Working adaptive favicon

wk9874 committed:
Rebuild

Commit:
chore: update build output [skip ci]

wk9874 committed:
Prettier

wk9874 committed:
Merge branch 'samueljackson92/improved-time-series-interface' of github.com:ukaea/toktagger into samueljackson92/improved-time-series-interface

Commit:
chore: update build output [skip ci]

samueljackson92 committed:
Merge branch 'main' into samueljackson92/improved-time-series-interface
